File: 2022/ds_1/day_6/valid-anagram.py
# https://leetcode.com/problems/valid-anagram/submissions/

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        store = [0] * 26
        aKey = ord('a')

        for i, j in s, t:
            store[ord(i) - aKey] += 1
            store[ord(j) - aKey] -= 1
        for i in store:
            if i != 0:
                return False
        return True

File: 2023/array/valid-anagram.js
// https://leetcode.com/problems/valid-anagram/description/

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function(s, t) {
    if (s.length !== t.length) return false;

    const storeS = new Array(26).fill(0)
    const storeT = new Array(26).fill(0)
    for(let i = 0; i < s.length; i++) {
        storeS[getCharNumber(s[i])]++;
        storeT[getCharNumber(t[i])]++;
    }
    return !storeS.some((_, i) => storeS[i] !== storeT[i])
};

function getCharNumber(char) {
    return char.charCodeAt(0) - 'a'.charCodeAt(0)
}